pkg/api/validation/validation.go

```diff
@@ -3972,9 +3972,10 @@ func validateEndpointSubsets(subsets []api.EndpointSubset, oldSubsets []api.Endp
 			//TODO: consider adding a RequiredOneOf() error for this and similar cases
 			allErrs = append(allErrs, field.Required(idxPath, "must specify `addresses` or `notReadyAddresses`"))
 		}
-		if len(ss.Ports) == 0 {
-			allErrs = append(allErrs, field.Required(idxPath.Child("ports"), ""))
-		}
+		// for headless service, we allow its endpoints not to have ports.
+		//if len(ss.Ports) == 0 {
```

Comment:
I don't know if it's clear enough when there is only a single comment without context...